<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro">

    <xacro:macro name="manipulator" params="parent">
        
        <!-- Arm Base -->
        <link name="arm_base_link">
             <visual>
                <geometry>
                    <box size="0.1 0.1 0.05"/>
                </geometry>
                <origin xyz="0 0 0.025" rpy="0 0 0"/>
                <material name="grey"/>
            </visual>
            <collision>
                <geometry>
                    <box size="0.1 0.1 0.05"/>
                </geometry>
                <origin xyz="0 0 0.025" rpy="0 0 0"/>
            </collision>
            <xacro:box_inertia m="0.5" l="0.1" w="0.1" h="0.05" xyz="0 0 0.025" rpy="0 0 0"/>
        </link>

        <joint name="arm_base_joint" type="fixed">
            <parent link="${parent}"/>
            <child link="arm_base_link"/>
            <origin xyz="0 0 0.2" rpy="0 0 0"/>
        </joint>

        <!-- Shoulder (Joint 1) - Rotates around Z -->
        <link name="shoulder_link">
            <visual>
                <geometry>
                    <cylinder radius="0.04" length="0.1"/>
                </geometry>
                <origin xyz="0 0 0.05" rpy="0 0 0"/>
                <material name="white"/>
            </visual>
            <collision>
                <geometry>
                    <cylinder radius="0.04" length="0.1"/>
                </geometry>
                <origin xyz="0 0 0.05" rpy="0 0 0"/>
            </collision>
             <xacro:cylinder_inertia m="0.3" r="0.04" h="0.1" xyz="0 0 0.05" rpy="0 0 0"/>
        </link>

        <joint name="shoulder_joint" type="revolute">
            <parent link="arm_base_link"/>
            <child link="shoulder_link"/>
            <origin xyz="0 0 0.05" rpy="0 0 0"/>
            <axis xyz="0 0 1"/>
            <limit lower="-3.14" upper="3.14" effort="10.0" velocity="1.0"/>
            <dynamics damping="0.1" friction="0.1"/>
        </joint>

        <!-- Upper Arm (Joint 2) - Rotates around Y -->
        <link name="upper_arm_link">
             <visual>
                <geometry>
                    <box size="0.3 0.05 0.05"/>
                </geometry>
                <origin xyz="0.15 0 0" rpy="0 0 0"/>
                <material name="blue"/>
            </visual>
            <collision>
                 <geometry>
                    <box size="0.3 0.05 0.05"/>
                </geometry>
                <origin xyz="0.15 0 0" rpy="0 0 0"/>
            </collision>
            <xacro:box_inertia m="0.5" l="0.3" w="0.05" h="0.05" xyz="0.15 0 0" rpy="0 0 0"/>
        </link>

        <joint name="elbow_joint" type="revolute"> <!-- Naming conventions: this is usually shoulder_lift or similar, but simplified here -->
            <parent link="shoulder_link"/>
            <child link="upper_arm_link"/>
            <origin xyz="0 0 0.1" rpy="0 0 0"/>
            <axis xyz="0 1 0"/>
            <limit lower="-2.0" upper="2.0" effort="10.0" velocity="1.0"/>
            <dynamics damping="0.1" friction="0.1"/>
        </joint>

        <!-- Forearm (Joint 3) - Rotates around Y -->
        <link name="forearm_link">
             <visual>
                <geometry>
                    <box size="0.25 0.04 0.04"/>
                </geometry>
                <origin xyz="0.125 0 0" rpy="0 0 0"/>
                <material name="black"/>
            </visual>
             <collision>
                <geometry>
                    <box size="0.25 0.04 0.04"/>
                </geometry>
                <origin xyz="0.125 0 0" rpy="0 0 0"/>
            </collision>
            <xacro:box_inertia m="0.4" l="0.25" w="0.04" h="0.04" xyz="0.125 0 0" rpy="0 0 0"/>
        </link>

        <joint name="wrist_joint" type="revolute">
            <parent link="upper_arm_link"/>
            <child link="forearm_link"/>
            <origin xyz="0.3 0 0" rpy="0 0 0"/>
            <axis xyz="0 1 0"/>
            <limit lower="-2.5" upper="2.5" effort="10.0" velocity="1.0"/>
            <dynamics damping="0.1" friction="0.1"/>
        </joint>
        
        <!-- Gripper Base -->
        <link name="gripper_base_link">
             <visual>
                <geometry>
                    <box size="0.05 0.08 0.05"/>
                </geometry>
                <origin xyz="0.025 0 0" rpy="0 0 0"/>
                <material name="grey"/>
            </visual>
            <collision>
                <geometry>
                    <box size="0.05 0.08 0.05"/>
                </geometry>
                <origin xyz="0.025 0 0" rpy="0 0 0"/>
            </collision>
             <xacro:box_inertia m="0.1" l="0.05" w="0.08" h="0.05" xyz="0.025 0 0" rpy="0 0 0"/>
        </link>

         <joint name="gripper_base_joint" type="fixed">
            <parent link="forearm_link"/>
            <child link="gripper_base_link"/>
            <origin xyz="0.25 0 0" rpy="0 0 0"/>
        </joint>

        <!-- Gripper Finger Left -->
        <link name="finger_left">
            <visual>
                <geometry>
                    <box size="0.08 0.01 0.03"/>
                </geometry>
                <origin xyz="0.04 0 0" rpy="0 0 0"/>
                <material name="red"/>
            </visual>
            <collision>
                <geometry>
                    <box size="0.08 0.01 0.03"/>
                </geometry>
                <origin xyz="0.04 0 0" rpy="0 0 0"/>
                <!-- High friction for grasping -->
                <surface>
                    <friction>
                        <ode>
                             <mu>1.0</mu>
                             <mu2>1.0</mu2>
                        </ode>
                    </friction>
                </surface>
            </collision>
            <xacro:box_inertia m="0.05" l="0.08" w="0.01" h="0.03" xyz="0.04 0 0" rpy="0 0 0"/>
        </link>

        <joint name="finger_left_joint" type="prismatic">
            <parent link="gripper_base_link"/>
            <child link="finger_left"/>
            <origin xyz="0.05 0.03 0" rpy="0 0 0"/>
            <axis xyz="0 1 0"/>
            <limit lower="-0.01" upper="0.04" effort="10.0" velocity="0.5"/>
             <dynamics damping="0.1" friction="0.1"/>
        </joint>

        <!-- Gripper Finger Right -->
        <link name="finger_right">
            <visual>
                <geometry>
                    <box size="0.08 0.01 0.03"/>
                </geometry>
                <origin xyz="0.04 0 0" rpy="0 0 0"/>
                <material name="red"/>
            </visual>
             <collision>
                <geometry>
                    <box size="0.08 0.01 0.03"/>
                </geometry>
                <origin xyz="0.04 0 0" rpy="0 0 0"/>
                 <surface>
                    <friction>
                        <ode>
                             <mu>1.0</mu>
                             <mu2>1.0</mu2>
                        </ode>
                    </friction>
                </surface>
            </collision>
            <xacro:box_inertia m="0.05" l="0.08" w="0.01" h="0.03" xyz="0.04 0 0" rpy="0 0 0"/>
        </link>

        <joint name="finger_right_joint" type="prismatic">
            <parent link="gripper_base_link"/>
            <child link="finger_right"/>
            <origin xyz="0.05 -0.03 0" rpy="0 0 0"/>
            <axis xyz="0 -1 0"/>
             <!-- Mimic left finger -->
            <mimic joint="finger_left_joint" multiplier="1" offset="0"/>
            <limit lower="-0.01" upper="0.04" effort="10.0" velocity="0.5"/>
             <dynamics damping="0.1" friction="0.1"/>
        </joint>
        
        <!-- Gazebo Joint Controller Plugins -->
        <gazebo>
            <!-- Joint Position Controllers for Arm -->
            <plugin filename="gz-sim-joint-position-controller-system" name="gz::sim::systems::JointPositionController">
                <joint_name>shoulder_joint</joint_name>
                <p_gain>100</p_gain>
                <i_gain>0.1</i_gain>
                <d_gain>1.0</d_gain>
                <i_max>1</i_max>
                <i_min>-1</i_min>
                <cmd_max>50</cmd_max>
                <cmd_min>-50</cmd_min>
            </plugin>
             <plugin filename="gz-sim-joint-position-controller-system" name="gz::sim::systems::JointPositionController">
                <joint_name>elbow_joint</joint_name>
                <p_gain>500</p_gain>
                <i_gain>10</i_gain>
                <d_gain>10.0</d_gain>
            </plugin>
             <plugin filename="gz-sim-joint-position-controller-system" name="gz::sim::systems::JointPositionController">
                <joint_name>wrist_joint</joint_name>
                <p_gain>100</p_gain>
                <i_gain>0.1</i_gain>
                <d_gain>1.0</d_gain>
            </plugin>
            
            <!-- Gripper Controller -->
            <plugin filename="gz-sim-joint-position-controller-system" name="gz::sim::systems::JointPositionController">
                <joint_name>finger_left_joint</joint_name>
                 <p_gain>500</p_gain>
                <i_gain>10</i_gain>
                <d_gain>10.0</d_gain>
            </plugin>
            <plugin filename="gz-sim-joint-position-controller-system" name="gz::sim::systems::JointPositionController">
                <joint_name>finger_right_joint</joint_name>
                 <p_gain>500</p_gain>
                <i_gain>10</i_gain>
                <d_gain>10.0</d_gain>
            </plugin>
        </gazebo>

    </xacro:macro>

</robot>
